File: mom_tp/mom.c
#define _XOPEN_SOURCE 700

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include <unistd.h>

#include "mom.h"
#include "libs/msq.h"
#include "mom_general.h"


// ------------------- PRIVATE FUNCTIONS ---------------------------


void fill_message(mom_message_t* m, mom_t* mom, opcode_t opcode, char* topic, char* payload) {
	m->mtype = mom->local_id;
	m->opcode = opcode;
	m->global_id = mom->global_id;
	m->local_id = mom->local_id;
	if(topic)	strcpy(m->topic, topic);
	if(payload)	strcpy(m->payload, payload);
}

void send_message_to_daemon(mom_t* mom, mom_message_t m, mom_message_t* response) {
	// Send m
	msq_send(mom->msqid_requester, &m, sizeof(mom_message_t));
	// Receive answer and store into response
	msq_rcv(mom->msqid_responser, response, sizeof(mom_message_t), mom->local_id);
}

bool topic_is_valid(char* topic) {
	if(strlen(topic) == 0) {
		printf("%d: Error, topic received has zero length!\n", getpid());
		return false;
	}
	if(strlen(topic) > TOPIC_LENGTH) {
		printf("%d: Error, topic cannot exceed lenght %d!\n", getpid(), TOPIC_LENGTH);
		return false;
	}
	for(int i = 0; i < strlen(topic); i++) {
		if((topic[i] == ' ') || (topic[i] == '.') || (topic[i] == ':') || (topic[i] == '!')) {
			printf("%d: Error, topic cannot have any space, . , : or !\n", getpid());
			return false;
		}
	}
	return true;
}


// --------------- LIBRARY PUBLIC FUNCTIONS ------------------------

mom_t* mom_create(){
	mom_t* mom = malloc(sizeof(mom_t));
	if(!mom)	return NULL;
	
	mom->local_id = getpid(); // Genius
	
	mom->global_id = mom->local_id;
	
	// Retrieve dameon queues' ids
	mom->msqid_requester = msq_create(QUEUE_REQUESTER);
	mom->msqid_responser = msq_create(QUEUE_RESPONSER);
	// Write and read create message to retrieve mom->global_id
	mom_message_t m = {0}, response = {0};
	fill_message(&m, mom, OC_CREATE, NULL, NULL);
	send_message_to_daemon(mom, m, &response);
	// Retrieve mom->global_id from response
	if((response.opcode != OC_ACK_SUCCESS) && (response.opcode != OC_ACK_FAILURE)) {
		printf("%d: MOM CRITICAL ON CREATING: Daemon answer was not ACK!\n", getpid());
		free(mom);
		return NULL;
	}
	if(response.opcode == OC_ACK_FAILURE) {
		printf("%d: Error creating mom, could not register on broker\n", getpid());
		free(mom);
		return NULL;
	}
	mom->global_id = response.global_id;
	return mom;
}

bool mom_publish(mom_t* mom, char* topic, const void *msg){
	if((!mom) || (!topic) || (!msg)) {
		printf("%d: Invalid argument in mom_publish: NULL received\n", getpid());
		return false;
	}
	if(!topic_is_valid(topic))	return false;
	mom_message_t m = {0}, response = {0};
	// Write and read publish message
	fill_message(&m, mom, OC_PUBLISH, topic, (char*) msg);
	send_message_to_daemon(mom, m, &response);
	if((response.opcode != OC_ACK_SUCCESS) && (response.opcode != OC_ACK_FAILURE)) {
		printf("%d: MOM CRITICAL ON PUBLISHING: Daemon answer was not ACK!\n", getpid());
		return false;
	}
	
	return response.opcode == OC_ACK_SUCCESS;
}

bool mom_subscribe(mom_t* mom, char* topic){
	if((!mom) || (!topic)) {
		printf("%d: Invalid argument in mom_publish: NULL received\n", getpid());
		return false;
	}
	if(!topic_is_valid(topic))	return false;
	mom_message_t m = {0}, response = {0};
	// Write and read subscribe message
	fill_message(&m, mom, OC_SUBSCRIBE, topic, NULL);
	send_message_to_daemon(mom, m, &response);
	if((response.opcode != OC_ACK_SUCCESS) && (response.opcode != OC_ACK_FAILURE)) {
		printf("%d: MOM CRITICAL ON SUBSCRIBING: Daemon answer was not ACK!\n", getpid());
		return false;
	}
	
	return response.opcode == OC_ACK_SUCCESS;
}

bool mom_receive(mom_t* mom, void* msg){
	if((!mom) || (!msg)) {
		printf("%d: Invalid argument in mom_publish: NULL received\n", getpid());
		return false;
	}
	// Locally receive message from daemon requester
	mom_message_t response = {0};
	msq_rcv(mom->msqid_responser, &response, sizeof(mom_message_t), mom->global_id);
	if(response.opcode != OC_DELIVERED) {
		printf("%d: MOM CRITICAL ON RECEIVING: Daemon has not delivered message!\n", getpid());
		return false;
	}
	// Copy received payload into msg
	strcpy((char*) msg, (char*) response.payload);
	return true;
}


void mom_destroy(mom_t* mom){
	if(!mom)	return;
	// Send destroy message
	mom_message_t m = {0}, response = {0};
	fill_message(&m, mom, OC_DESTROY, NULL, NULL);
	send_message_to_daemon(mom, m, &response);
	if((response.opcode != OC_ACK_SUCCESS) && (response.opcode != OC_ACK_FAILURE)) 
		printf("%d: MOM CRITICAL ON DESTROYING: Daemon answer was not ACK!\n", getpid());
	
	// Release shared memory nonetheless
	free(mom);
	// Must not destroy mom->msqids since they're mom_daemon queues
}

File: mom_tp/mom.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import sysv_ipc as sv
import os
import struct

# --------------------- Python MOM API --------------------- 



# ----------------------- Constants ------------------------

OC_CREATE = 1
OC_DESTROY = 2 
OC_PUBLISH = 3
OC_SUBSCRIBE = 4
OC_ACK_SUCCESS = 5
OC_ACK_FAILURE = 6
OC_DELIVERED = 7
OC_SEPPUKU = 8

PAYLOAD_SIZE = 255
TOPIC_LENGTH = 100

QUEUE_RESPONSER = "queue_resp.temp"
QUEUE_REQUESTER = "queue_req.temp"

MSQ_MAGIC_INT = 11	# For ftok

# --------------- Auxiliar MomMessage class ----------------

'''
Format of mom_message_t as declared in mom_general.h:

	long mtype;
	
	long local_id;
	long global_id;
	opcode_t opcode;
	char topic[TOPIC_LENGTH];
	
	char payload[PAYLOAD_SIZE];
	
'''

class MomMessage:
	def __init__(self, localId=0, globalId=0, opcode=0):
		self.localId = localId
		self.globalId = globalId
		self.opcode  = opcode
		self.topic = "".ljust(TOPIC_LENGTH, "\0")
		self.payload = "".ljust(PAYLOAD_SIZE, "\0")
	
	def setTopic(self, topic):
		self.topic = topic.ljust(TOPIC_LENGTH, "\0")
		
	def setPayload(self, payload):
		self.payload = payload.ljust(PAYLOAD_SIZE, "\0")
	
	def getStringRepresentation(self):
		# Pack the message!
		r = struct.pack(b'@lli', self.localId, self.globalId, self.opcode)
		r = r + self.topic + self.payload
		return r
	
	def setFromStringRepresentation(self, s):
		TEXT_START = 12
		(self.localId, self.globalId, self.opcode) = struct.unpack(b'@lli', s[0:TEXT_START])
		self.topic = s[TEXT_START:TEXT_START+TOPIC_LENGTH]
		self.payload = s[TEXT_START+TOPIC_LENGTH:TEXT_START+TOPIC_LENGTH+PAYLOAD_SIZE]
	
	def getGlobalId(self):
		return self.globalId
		
	def getOpcode(self):
		return self.opcode
		
	def getPayload(self):
		return self.payload
		
	def printMessage(self):
		oc_str = ["", "CREATE", "DESTROY", "PUBLISH", "SUBSCRIBE", "ACK_SUCCESS", "ACK_FAILURE", "DELIVERED", "SEPPUKU"]
		print("Local id: " + str(self.localId))
		print("Global id: " + str(self.globalId))
		print("Opcode: " + oc_str[self.opcode])
		print("Topic: " + self.topic + ".")
		print("Payload: " + self.payload + ".")

# ----------- The fantabolously magic Mom class -----------

class Mom:
	def __init__(self):
		self.localId = os.getpid()
		self.globalId = os.getpid()
		
		# Retrieve msqs
		keyReq = sv.ftok(QUEUE_REQUESTER, MSQ_MAGIC_INT, True)
		keyResp = sv.ftok(QUEUE_RESPONSER, MSQ_MAGIC_INT, True)
		self.msqReq = sv.MessageQueue(keyReq, flags=sv.IPC_CREAT, mode=0644)
		self.msqResp = sv.MessageQueue(keyResp, flags=sv.IPC_CREAT, mode=0644)
		
		# Retrieve globalId
		m = MomMessage(self.localId, self.globalId, OC_CREATE)
		response = self._sendMessageToDaemon(m)
		ropc = response.getOpcode()
		if (ropc != OC_ACK_FAILURE) and (ropc != OC_ACK_SUCCESS):
			print(str(os.getpid()) + ": MOM CRITICAL ON CREATING: Daemon answer was not ACK!\n")
			self = None
			return
		if ropc == OC_ACK_FAILURE:
			print(str(os.getpid()) + ": Error creating mom, could not register on broker\n")
			self = None
			return
		self.globalId = response.getGlobalId()
		
	def _topicIsValid(self, topic):
		if(len(topic) == 0):
			printf(str(os.getpid()) + ": Error, topic received has zero length!\n", )
			return False;
		
		if(len(topic) > TOPIC_LENGTH):
			printf(str(os.getpid()) + ": Error, topic cannot exceed lenght: " + str(TOPIC_LENGTH) + "!\n")
			return False;
		
		for c in topic:
			if((c == ' ') or (c == '.') or (c == ':') or (c == '!')):
				printf(str(os.getpid()) + ": Error, topic cannot have any space, . , : or !\n")
				return False;

		return True;
	
	def _sendMessageToDaemon(self, m):
		# Send message
		self.msqReq.send(m.getStringRepresentation(), type=self.localId)
		# Await response
		r, _ = self.msqResp.receive(type=self.localId)
		response = MomMessage()
		response.setFromStringRepresentation(r)
		return response
	
	def publish(self, topic, msg):
		if (topic is None) or (msg is None):
			print(str(os.getpid()) + ": Invalid argument in mom_publish: None received\n", );
			return False
		if not self._topicIsValid(topic):
			return False
		# Write and read publish message
		m = MomMessage(self.localId, self.globalId, OC_PUBLISH)
		m.setTopic(topic)
		m.setPayload(msg)
		response = self._sendMessageToDaemon(m)
		ropc = response.getOpcode()
		if (ropc != OC_ACK_FAILURE) and (ropc != OC_ACK_SUCCESS):
			print(str(os.getpid()) + ": MOM CRITICAL ON PUBLISHING: Daemon answer was not ACK!\n")
			self = None
			return

		return (ropc == OC_ACK_SUCCESS)
		
	def receive(self):
		# Locally receive message from daemon requester
		r, _ = self.msqResp.receive(type=self.globalId)
		response = MomMessage()
		response.setFromStringRepresentation(r)
		ropc = response.getOpcode()
		if (ropc != OC_DELIVERED):
			print(str(os.getpid()) + ": MOM CRITICAL ON RECEIVING: Daemon has not delivered message!\n")
			return None
		return response.getPayload()
		
	def subscribe(self, topic):
		if (topic is None):
			print(str(os.getpid()) + ": Invalid argument in mom_subscribe: None received\n", );
			return False
		if not self._topicIsValid(topic):
			return False
		# Write and read subscribe message
		m = MomMessage(self.localId, self.globalId, OC_SUBSCRIBE)
		m.setTopic(topic)
		response = self._sendMessageToDaemon(m)
		ropc = response.getOpcode()
		if (ropc != OC_ACK_FAILURE) and (ropc != OC_ACK_SUCCESS):
			print(str(os.getpid()) + ": MOM CRITICAL ON SUBSCRIBING: Daemon answer was not ACK!\n")
			self = None
			return

		return (ropc == OC_ACK_SUCCESS)
	
	def __del__(self):
		# Send destroy message
		m = MomMessage(self.localId, self.globalId, OC_DESTROY)
		response = self._sendMessageToDaemon(m)
